pdfmaker_preprocessing.rb

```diff
@@ -70,144 +194,135 @@ def writeImageErrors(arr, file)
-    ls = ftp.nlst()
-    unless ls.include?(parentfolder)
-      ftp.mkdir(parentfolder)
-    end
-    ftp.chdir(parentfolder)
-    ls = ftp.nlst()
-    unless ls.include?(childfolder)
-      ftp.mkdir(childfolder)
-    end
-    ftp.chdir(childfolder)
-    files = ftp.nlst()
-    ftp.close
-    return files
-  end
+#wrapping ftp class method in its own separate method: to write to jsonlog here and leave class methods more general.
+def ftpUpload(ftpdirint, pdftmp_dir, ftplist, logkey='')
+  ftpstatus = Ftpfunctions.uploadImg(ftpdirint, pdftmp_dir, ftplist)
+  return ftpstatus
+rescue => logstring
+  return ''
+ensure
+  Mcmlln::Tools.logtoJson(@log_hash, logkey, logstring)
+end
 
-  def self.uploadImg(dir, srcdir, filenames)
-    ftp = Ftpfunctions.loginFTP(@@ftp_url, @@ftp_username, @@ftp_password)
-    files = ftp.chdir(dir)
-    filenames.each do |p|
-      filepath = File.join(srcdir, p)
-      ftp.putbinaryfile(filepath)
-    end
-    files = ftp.nlst()
-    ftp.close
-    return files
-  end
+## wrapping Bkmkr::Tools.runnode in a new method for this script; to return a result for json_logfile
+def localRunNode(jsfile, args, logkey='')
+	Bkmkr::Tools.runnode(jsfile, args)
+rescue => logstring
+ensure
+  Mcmlln::Tools.logtoJson(@log_hash, logkey, logstring)
+end
+
+# fixes images in html, keep final words and ellipses from breaking
+def fixHtmlImageSrcAndKTs(pdf_tmp_html, ftpdirext, logkey='')
+  # .gsub(/([a-zA-Z0-9][a-zA-Z0-9][a-zA-Z0-9]\s\. \. \.)/,"<span class=\"bookmakerkeeptogetherkt\">\\0</span>")
+  filecontents = File.read(pdf_tmp_html).gsub(/src="images\//,"src=\"#{ftpdirext}/")
+                                        .gsub(/([a-zA-Z0-9]?[a-zA-Z0-9]?[a-zA-Z0-9]?\s\. \. \.)/,"<span class=\"bookmakerkeeptogetherkt\">\\0</span>")
+                                        .gsub(/(\s)(\w\w\w*?\.)(<\/p>)/,"\\1<span class=\"bookmakerkeeptogetherkt\">\\2</span>\\3")
+  return filecontents
+rescue => logstring
+  return ''
+ensure
+  Mcmlln::Tools.logtoJson(@log_hash, logkey, logstring)
+end
+
+
+# fixes em dash breaks (requires UTF 8 encoding)
+def fixEmdashes(pdf_tmp_html, logkey='')
+  filecontents = File.read(pdf_tmp_html, :encoding=>"UTF-8").gsub(/(.)?(—\??\.?!?”?’?)(.)?/,"\\1\\2&\#8203;\\3")
+                                                            .gsub(/(<p class="FrontSalesQuotefsq">“)(A)/,"\\1&\#8202;\\2")
+  return filecontents
+rescue => logstring
+  return ''
+ensure
+  Mcmlln::Tools.logtoJson(@log_hash, logkey, logstring)
 end
 
+
 # ---------------------- PROCESSES
+data_hash = readConfigJson('read_config_json')
+#local definition(s) based on config.json
+project_dir = data_hash['project']
+stage_dir = data_hash['stage']
+# including: ftp url
+ftpdirext = "http://www.macmillan.tools.vhost.zerolag.com/bookmaker/bookmakerimg/#{project_dir}_#{stage_dir}/#{Metadata.pisbn}"
+ftpdirint = "/files/html/bookmaker/bookmakerimg/#{project_dir}_#{stage_dir}/#{Metadata.pisbn}"
```

Comment:
Those are all dependent on reading the config.rb in line 278, which has to come after the "readConfigJson" method..  :(  ?